servicetalk-loadbalancer-experimental/src/main/java/io/servicetalk/loadbalancer/RoundRobinSelector.java

```diff
@@ -134,22 +137,41 @@ private static Scheduler buildScheduler(AtomicInteger index, List<? extends Host
     }
 
     private abstract static class Scheduler {
+
+        private final AtomicInteger index;
+        protected final int hostsSize;
+
+        Scheduler(final AtomicInteger index, final int hostsSize) {
+            this.index = index;
+            this.hostsSize = hostsSize;
+        }
+
+        // Get the index of the next host
         abstract int nextHost();
+
+        protected final int nextIndex() {
+            return index.getAndIncrement();
+        }
+
+        // Let the scheduler know the index was found to be unhealthy in an attempt to avoid causing the node
+        // after an unhealthy node to effectively receive double traffic.
+        final void foundUnhealthy(int index) {
+            int i = this.index.get();
```

Comment:
Note that the passed `index` went through `Integer.toUnsignedLong` conversion, while `this.index` can be negative. Consider adding a test that starts from a state with negative `this.index`.

Reply:
Exceptional catch. 😃 
I added some tests as you suggested.